File: app/src/main/java/org/tesira/civic/HomeSpecialsAdapter.java
package org.tesira.civic;

import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.tesira.civic.db.CivicViewModel;

import java.util.List;

/**
 * Adapter for showing a list of Special Abilities or Immunities against on the dashboard.
 */
public class HomeSpecialsAdapter extends RecyclerView.Adapter<HomeSpecialsAdapter.ViewHolder> {

    private String[] localDataSet = new String[0];
    private final CivicViewModel viewModel;

    /**
     * Provide a reference to the type of views that you are using
     * (custom ViewHolder).
     */
    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView textView;

        public ViewHolder(View view) {
            super(view);
            // Define click listener for the ViewHolder's View

            textView = view.findViewById(R.id.textView);
        }

        public TextView getTextView() {
            return textView;
        }
    }

    public HomeSpecialsAdapter(CivicViewModel viewModel) {
        this.viewModel = viewModel;
        this.localDataSet = new String[0]; // initial leer
    }

    // Create new views (invoked by the layout manager)
    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        // Create a new view, which defines the UI of the list item
        View view = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.item_row_specials, viewGroup, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, final int position) {
        viewHolder.getTextView().setText(localDataSet[position]);
        viewHolder.getTextView().setBackgroundResource(R.drawable.specials_background);
        if (viewHolder.getTextView().getText().toString().startsWith("_")) {
            viewHolder.getTextView().setTypeface(viewHolder.getTextView().getTypeface(), Typeface.BOLD);
        }
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return localDataSet.length;
    }

    public void submitSpecialsList(List<String> newSpecialsAndImmunities) {
        if (newSpecialsAndImmunities != null) {
            this.localDataSet = newSpecialsAndImmunities.toArray(new String[0]);
        } else {
            this.localDataSet = new String[0];
        }
        notifyDataSetChanged();
    }

}

File: app/src/main/kotlin/org/tesira/civic/HomeSpecialsAdapter.kt
package org.tesira.civic

import android.annotation.SuppressLint
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

/**
 * Adapter for showing a list of Special Abilities or Immunities against on the dashboard.
 */
class HomeSpecialsAdapter : RecyclerView.Adapter<HomeSpecialsAdapter.ViewHolder?>() {
    private var localDataSet: MutableList<String> = ArrayList<String>()

    /**
     * Provide a reference to the type of views that you are using
     * (custom ViewHolder).
     */
    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val textView: TextView? = view.findViewById<TextView?>(R.id.textView)
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
        // Create a new view, which defines the UI of the list item
        val view = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.item_row_specials, viewGroup, false)

        return ViewHolder(view)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.textView!!.text = localDataSet[position]
        viewHolder.textView.setBackgroundResource(R.drawable.bg_specials)
        if (viewHolder.textView.text.toString().startsWith("_")) {
            viewHolder.textView.setTypeface(viewHolder.textView.typeface, Typeface.BOLD)
        }
    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {
        return localDataSet.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun submitSpecialsList(newSpecialsAndImmunities: List<String>) {
        this.localDataSet.clear()
        this.localDataSet.addAll(newSpecialsAndImmunities)
        notifyDataSetChanged()
    }
}